File: product_sum/product_sum.java
import java.util.List;

public class Main {

    public static int productSum(List<Object> array, int depth) {
        int sum = 0;
        
        for (Object item : array) {
            if (item instanceof List<?>) {
                sum += productSum((List<Object>) item, depth + 1);
            } else {
                sum += (int) item;
            }
        }

        return sum * depth;
    }

    public static void main(String[] args) {
        List<Object> array = List.of(1, List.of(2, 1));
        System.out.println(productSum(array, 1));  
    }
}

File: product_sum/product_sum.py
# O(n) time | O(d) space where n is the length of the array and d is the depth
def product_sum(array, depth = 1):
  sum = 0
  for item in array:
    if type(item) == list:
      sum += product_sum(item, depth + 1)
    else:
      sum += item
  return sum * depth

print(product_sum([1, [2, 1]]))